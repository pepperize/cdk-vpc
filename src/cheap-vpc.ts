import { aws_ec2 } from "aws-cdk-lib";
import { VpcProps } from "aws-cdk-lib/aws-ec2/lib/vpc";
import { Construct } from "constructs";

export class CheapVpc extends aws_ec2.Vpc {
  public constructor(scope: Construct, id: string, props?: VpcProps) {
    const natInstanceProvider = aws_ec2.NatProvider.instance({
      instanceType: aws_ec2.InstanceType.of(aws_ec2.InstanceClass.T3, aws_ec2.InstanceSize.NANO),
    });

    super(scope, id, {
      natGatewayProvider: natInstanceProvider,
      natGateways: 1,
      maxAzs: 2,
      ...props,
    });
  }
}
