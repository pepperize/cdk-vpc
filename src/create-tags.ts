import { Stack } from "aws-cdk-lib";
import { makeUniqueId } from "aws-cdk-lib/core/lib/private/uniqueid";
import { AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId } from "aws-cdk-lib/custom-resources";
import { Construct } from "constructs";

export interface Tag {
  /**
   * The key of the tag.
   *
   * Constraints: Tag keys are case-sensitive and accept a maximum of 127 Unicode characters. May not begin with aws:.
   */
  readonly key: string;
  /**
   * The value of the tag.
   *
   * Constraints: Tag values are case-sensitive and accept a maximum of 256 Unicode characters.
   */
  readonly value: string;
}

/**
 * Adds or overwrites only the specified tags for the specified Amazon EC2 resource or resources. When you specify an existing tag key, the value is overwritten with the new value. Each resource can have a maximum of 50 tags. Each tag consists of a key and optional value. Tag keys must be unique per resource.
 *
 * @see https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/Using_Tags.html
 * @see https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateTags.html
 */
export interface CreateTagsProps {
  /**
   * The IDs of the ec2 resources, separated by spaces. Constraints: Up to 1000 resource IDs. We recommend breaking up this request into smaller batches.
   *
   * @see https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/Using_Tags.html#tag-resources
   */
  readonly resourceIds: string[];

  /**
   * The tags. The value parameter is required, but if you don't want the tag to have a value, specify the parameter with no value, and we set the value to an empty string.
   */
  readonly tags: Tag[];
}

export class CreateTags extends Construct {
  public constructor(scope: Construct, id: string, props: CreateTagsProps) {
    super(scope, id);

    const { resourceIds, tags } = props;

    new AwsCustomResource(this, "CreateTags", {
      onCreate: {
        service: "EC2",
        action: "createTags", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/EC2.html#createTags-property
        region: Stack.of(this).region,
        parameters: {
          Resources: resourceIds,
          Tags: tags.map((tag) => {
            return { Key: tag.key, Value: tag.value };
          }),
        },
        physicalResourceId: PhysicalResourceId.of(makeUniqueId(resourceIds)),
      },
      installLatestAwsSdk: false,
      policy: AwsCustomResourcePolicy.fromSdkCalls({
        resources: AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
    });
  }
}
