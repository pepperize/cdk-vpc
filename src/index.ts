export * from "./cheap-vpc";
export * from "./create-tags";
