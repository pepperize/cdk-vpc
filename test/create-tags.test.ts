import { App, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { CreateTags } from "../src";

describe("PolicyAttachment", () => {
  it("Should match snapshot", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "Stack");
    new CreateTags(stack, "CreateTags", {
      resourceIds: ["subnet-0a4b64097b272baeD"],
      tags: [
        {
          key: "kubernetes.io/role/internal-elb",
          value: "1",
        },
      ],
    });

    // When
    const template = Template.fromStack(stack);

    // Then
    expect(template).toMatchSnapshot();
  });
});
