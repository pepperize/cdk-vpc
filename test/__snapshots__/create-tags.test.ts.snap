// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PolicyAttachment Should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "11d371859e49f3b1a49896583254890801a40beada56e4c84409a179a221e2e2.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CreateTags59113D99": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "CreateTagsCustomResourcePolicy69E92226",
      ],
      "Properties": Object {
        "Create": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"EC2\\",\\"action\\":\\"createTags\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"parameters\\":{\\"Resources\\":[\\"subnet-0a4b64097b272baeD\\"],\\"Tags\\":[{\\"Key\\":\\"kubernetes.io/role/internal-elb\\",\\"Value\\":\\"1\\"}]},\\"physicalResourceId\\":{\\"id\\":\\"subnet0a4b64097b272baeD\\"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": false,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "CreateTagsCustomResourcePolicy69E92226": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ec2:CreateTags",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CreateTagsCustomResourcePolicy69E92226",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
